### PR TITLE
chore(broker): restore logstream

### LINKED ISSUE #2429: Restore log stream when events are missing
Since introducing the `DistributedLogstream` primitive, we now have two logs: the atomix raft log, which keeps track of operations performed on the various primitives in a raft group, and our own local `LogStream`, which are both truncated independently under different conditions.

It can happen that when a node has been gone for a while, that when it rejoins, both logs have been truncated on all nodes. On the Atomix side, to ensure that a node can rejoin even after truncation, a `Backup` operation is performed, which would normally store the state of the primitive up until a given position; this is unfortunately cost prohibitive in our case, as the state is the log data, which can be several gigabytes (and would thus exist in duplicate). Our own `RocksDB` snapshots are also not viable, as reprocessing still requires us to have missing events.

The proposed solution is thus to write the last committed position in the Atomix `Backup`, and during the `Restore` operation of the primitive, attempt to restore the local log to a state consistent with the other nodes of its raft by replicating any necessary events and/or `RocksDB` snapshots.

It could look like this:

1. Given a backup operation which has `commitPosition`, and a local log with latest commit position `x < commitPosition`:
  1. Determine who to restore from and what can be restored from them (#2424)
  1. Restore from that node what we can
      - Replicate any missing snapshot (truncate log if replicating from snapshot) (#2423)
      - Replicate any missing events from the remote log into the local log (#2425)
      - Provide a combined future to listen for failure/completion of all operations
  1.  Update state, and if latest local commit position is still less than `commitPosition`, start over; otherwise complete restore operation

### PR DESCRIPTION
**Pull request by deepthidevaki**
* When DistributedLogstream primitive detects that it is missing event because atomix raft gives a backup state, it restores logstream and stream processor snapshot so that it is in a consistent state. 
* Also moved to RaftRoleChangeListeners instead of LeaderElection primitive, because of possible bugs in atomix leader election listeners.
 
closes #2429

### CODE REVIEW

File: logstreams/src/main/java/io/zeebe/logstreams/impl/service/LogStreamService.java [hunk 1]
```diff
@@ -280,6 +284,21 @@ public void setExporterPositionSupplier(final Supplier<Long> supplier) {
...
+    restoring.set(true);
+  }
+
+  @Override
+  public void completeRestore() {
+    restoring.set(false);
+  }
+
+  @Override
+  public boolean canProcess() {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/impl/service/LogStreamService.java]: Let's add an uglier workaround in the `LogstreamConfig` as we will figure out a proper way to do this in #2484

**Review by npepinpe — CHANGES_REQUESTED** [review]
:+1:

**Review by npepinpe — APPROVED** [review]

**Comment by deepthidevaki** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors r+

not sure why bors always ignores my retry :thinking:

**Comment by npepinpe** [comment]
Merging manually as the builds fail due to different flaky tests.

**Pull request merged**